src/webgpu/api/validation/encoding/programmable/pipeline_bind_group_compat.spec.ts

```diff
@@ -14,8 +14,301 @@ TODO: subsume existing test, rewrite fixture as needed.
+          break;
+        }
+        case 'read-only-storage': {
+          decoration = 'storage, read';
+          break;
+        }
+        default:
+          unreachable();
+      }
+    }
+    declare = `var<${decoration}> buf: Buf`;
+  } else if (binding.externalTexture !== undefined) {
+    declare = `var tex : texture_external;`;
+  } else if (binding.sampler !== undefined) {
+    let suffix: string;
+    switch (binding.sampler.type) {
+      case 'filtering':
+      case 'non-filtering': {
+        suffix = 'sampler';
+        break;
+      }
+      case 'comparison': {
+        suffix = 'sampler_comparison';
+        break;
+      }
+      default:
+        unreachable();
+    }
+    declare = `var samp : ${suffix}`;
+  } else if (binding.storageTexture !== undefined) {
+    let suffix: string;
+    let textureType: string;
+    let texelFormat: string;
+
+    if (binding.storageTexture.viewDimension === undefined) {
+      textureType = 'texture_storage_2d';
+    } else {
+      switch (binding.storageTexture.viewDimension) {
+        case '1d': {
+          textureType = 'texture_storage_1d';
+          break;
+        }
+        case '2d': {
+          textureType = 'texture_storage_2d';
+          break;
+        }
+        case '2d-array':
+        case '3d': {
+          textureType = 'texture_storage_3d';
+          break;
+        }
+        default:
+          // Cannot be 'cube' or 'cube-array'
+          unreachable();
+      }
+    }
+
+    switch (binding.storageTexture.format) {
+      // float
+      case 'rgba8unorm':
+      case 'rgba8snorm':
+      case 'rgba16float':
+      case 'r32float':
+      case 'rg32float':
+      case 'rgba32float': {
+        texelFormat = 'f32';
+        break;
+      }
+
+      // uint
+      case 'rgba8uint':
+      case 'rgba16uint':
+      case 'r32uint':
+      case 'rg32uint':
+      case 'rgba32uint': {
+        texelFormat = 'u32';
+        break;
+      }
+
+      // sint
+      case 'rgba8sint':
+      case 'rgba16sint':
+      case 'r32sint':
+      case 'rg32sint':
+      case 'rgba32sint': {
+        texelFormat = 'i32';
+        break;
+      }
+      default:
+        unreachable();
+    }
+
+    suffix = `${textureType}<${texelFormat}`;
+
+    if (binding.storageTexture.access === 'write-only') {
+      suffix += ', write>';
+    } else {
+      suffix += '>';
+    }
+
+    declare = `var storeTex: ${suffix}`;
+  } else if (binding.texture !== undefined) {
+    let textureType: string;
+    let texelFormat: string = '';
+
+    const isMultisampled =
+      binding.texture.multisampled !== undefined && binding.texture.multisampled;
+    assert(
+      isMultisampled &&
+        (binding.texture.viewDimension === undefined || binding.texture.viewDimension === '2d')
+    );
+    let isDepth: boolean = false;
+
+    if (binding.texture.sampleType === undefined) {
+      texelFormat = 'f32';
+    } else {
+      switch (binding.texture.sampleType) {
```

Comment:
could be `switch (binding.texture.sampleType ?? 'float')` instead of having the extra `if ()` above

(`??` is the "null coalescing operator")